### PR TITLE
Rename reserved function names

### PR DESCRIPTION
**Pull request by pfl**
Rename reserved function names in drivers/ subdirectory. Update
function macros concatenatenating function names with '##'. As
there is a conflict between the existing gpio_sch_manage_callback()
and _gpio_sch_manage_callback() names, leave the latter unmodified.

While waiting for the prefix discussion to conclude, add also renaming for arch/.

The reserved function names are discussed in #9882

### CODE REVIEW

File: drivers/modem/wncm14a2a.c
Comment (mike-scott) [inline drivers/modem/wncm14a2a.c]: Why is this z_hexdump when the renaming was from _hexdump() -> hexdump()?

**Review by mike-scott — DISMISSED** [review]

File: drivers/modem/wncm14a2a.c
Comment (ceolin) [reply drivers/modem/wncm14a2a.c]: yeah, symbol is part of this driver so does not need have zephyr's namespace IMHO

Comment (mike-scott) [reply drivers/modem/wncm14a2a.c]: I meant `z_hexdump` doesn't exist.  `_hexdump` was renamed to `hexdump` (minus the underscore)

Comment (ceolin) [reply drivers/modem/wncm14a2a.c]: I see, my comment is still valid though

Comment (pfl) [reply drivers/modem/wncm14a2a.c]: Renaming accident, I'll fix it up to just hexdump().

File: include/kernel.h
Comment (jhedberg) [inline include/kernel.h]: This seems very unintuitive to me. This API operates on an object called k_timer, so whatever the prefix is (or isn't) the name should still contain K_TIMER

File: include/kernel.h
Comment (jhedberg) [inline include/kernel.h]: Same here. The object operated on is called k_fifo, so the name should contain K_FIFO

File: include/kernel.h
Comment (jhedberg) [inline include/kernel.h]: Same here.

**Review by jhedberg — DISMISSED** [review]

File: include/kernel.h
Comment (andrewboie) [reply include/kernel.h]: I disagree.
These are private APIs, this is why the name had a leading underscore. End users shouldn't be using this. The naming convention for private objects is to prefix with z_ instead of k_

Comment (tbursztyka) [reply include/kernel.h]: What about Z_K_TIMER ? Though it's a bit ugly, it does keep the full object type name. Knowing also that 'K' has a strong meaning, unlike Z... unless we rename all the kernel types k_<> to z_<> but that would be even weirder.

Comment (jhedberg) [reply include/kernel.h]: I'm not opposed to the z_, just that the name of the API should contain the object that it operates on (it's a k_timer not a z_timer).

Btw, what exactly do you mean by "internal"? That only the k_timer code should be using this? If that's the case, how do you propose getting rid of users like this: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/hci_core.c#L72

Comment (andrewboie) [reply include/kernel.h]: > I'm not opposed to the z_, just that the name of the API should contain the object that it operates on (it's a k_timer not a z_timer).

I don't really have a strong feeling about this one way or another, aesthetically z_k_ seems awkward but it's just a few macros. I'll go with the prevailing opinion.

> Btw, what exactly do you mean by "internal"? That only the k_timer code should be using this? 

I am not sure where our policy is documented on subsys/ with regards to private kernel APIs.

For those macros in particular, IIRC this was due to objects declared with them not being accessible to user mode.

The solution I'm currently pursuing is to implement futexes, they don't any restrictions where their memory can live.

Comment (jhedberg) [reply include/kernel.h]: Btw, either of these changes will likely increase the chance of application code accessing the internal APIs, since for an uneducated/naive developer who's looking at include/kernel.h a z_* API from a project called Zephyr will not look in anyway suspicious or communicate a "stay away!" message. I realize the ship may already have sailed, but I'd have at least considered putting any internal APIs into a separate `kernel_internal.h`, and if we want to be completely clear to the user throw an `_internal` suffix at them.

Comment (pfl) [reply include/kernel.h]: The leading underscore is the one targeted, any other prefix not starting with it is fine once we agree on it.

**Comment by pfl** [comment]
Added arch/ function renaming, as it anyway is based on the changes in drivers/.

File: arch/posix/include/posix_arch_internal.h
Comment (aescolar) [inline arch/posix/include/posix_arch_internal.h]: Please rename this PC_SAFE_CALL and pc_safe_call respectively (this macro and function are meant to be private to the posix arch, and should not appear to be usable by anybody else)

**Review by aescolar — DISMISSED** [review]

File: arch/posix/include/posix_arch_internal.h
Comment (pfl) [reply arch/posix/include/posix_arch_internal.h]: Fixed.

**Comment by pfl** [comment]
Fixed up reserved function names in boards, soc and subsys. The functions @jhe commented about are not yet renamed; these can in the worst case be renamed in a separate new patch in this PR.

File: ext/hal/ti/simplelink/source/ti/devices/cc32xx/inc/hw_gpio.h
Comment (vanti) [inline ext/hal/ti/simplelink/source/ti/devices/cc32xx/inc/hw_gpio.h]: I don't think this file needs to be changed. The comments are referring to some register name, not Zephyr function name.

**Review by vanti — DISMISSED** [review]

**Comment by pfl** [comment]
> 1: UC3 Title does not follow [subsystem]: [subject] (and should not start with literal subsys:): "subsys: Rename reserved function names"

Hrm, this commit follows the above pattern *exactly*...

File: ext/hal/ti/simplelink/source/ti/devices/cc32xx/inc/hw_gpio.h
Comment (pfl) [reply ext/hal/ti/simplelink/source/ti/devices/cc32xx/inc/hw_gpio.h]: Automatic name conversion bug, fixed.

Comment (vanti) [reply ext/hal/ti/simplelink/source/ti/devices/cc32xx/inc/hw_gpio.h]: LGTM

**Review by ioannisg — COMMENTED** [review]
OK for soc/arm/nordic_nrf changes

**Review by Laczen — APPROVED** [review]
ok for subsys/fs/nvs

**Review by vanwinkeljan — APPROVED** [review]
ok for display related changes

**Comment by pfl** [comment]
Updated to latest master.

File: include/kernel.h
Comment (ceolin) [reply include/kernel.h]: keep prefix `k` for private kernel API is ok for me. @pfl what do you think ? I can't disagree with @jhedberg points.

Comment (pfl) [reply include/kernel.h]: Fine with me, but you'd need to spell out if you want a subset of the former '_k_*' functions or all of them changed to prefix 'k_'. I'm going to assume all, so I'll update the patch set accordingly.

The difficult part of this discussion is that it is nowhere explicitly stated which functions are actually exported to the rest of the system ("kernel" and "user space"), and which are "kernel" internal. Johan's comment about include/ and kernel/include is spot on, even I wouldn't know the difference.

We probably need to document this and start moving include files into the same hierarchy we now have for ext/, subsys/ etc.. Right now include/ is a bit too messy when just trying to comprehend available features by reading the individual .h files and trying to figure out if they are ext/, subsys/ or something elsel. Going in the direction from Kconfig variables to finding the relevant include files is the way it seems to work now.

Comment (andrewboie) [reply include/kernel.h]: > keep prefix k for private kernel API is ok for me

No, please do not do this.
These private APIs, which can disappear or change at *any* time, need to be clearly distinguishable from Zephyr's public interfaces!
z_ makes this very clear that this is a different namespace.

**Comment by pfl** [comment]
Conflicts resolved, updated to latest master.

File: include/kernel.h
Comment (jhedberg) [reply include/kernel.h]: > These private APIs, which can disappear or change at any time, need to be clearly distinguishable from Zephyr's public interfaces!

@andrewboie I completely agree with this.

> z_ makes this very clear that this is a different namespace.

Indeed, but not that it's a private API. What makes things worse is that currently these are in the same public header file as the public APIs and (at least some) private APIs don't even have a code comment saying they're private. For a naïve reader the private APIs will not look private at all.

Comment (pfl) [reply include/kernel.h]: > What makes things worse is that currently these are in the same public header file as the public APIs and (at least some) private APIs don't even have a code comment saying they're private. For a naive reader the private APIs will not look private at all.

But that is then a bit of a different issue. I'm not happy at all that the private/public/subsys/ext header files are mixed into the same directory, at minimum the public and private APIs should be in different files and the files properly ordered into a hierarchy. That seems to be a task for 1.15, though.

Comment (jhedberg) [reply include/kernel.h]: @pfl understood, and agreed mostly. I just think that (ignoring the standard compliance, reserved namespace issue) the _ prefix had at least a slightly better chance of screaming "private!" than z_ has.

Comment (pfalcon) [reply include/kernel.h]: > Indeed, but not that it's a private API. What makes things worse is that currently these are in the same public header file as the public APIs

+1 for splitting public into private APIs into separate headers. That's the best way to keep namespace clean. Except that in our case it will resolve the situation only "so much", as e.g. we have private macros used to define public APIs. Still, it's the best mechanism we'd have to structure and organize Zephyr namespace.

Comment (pfl) [reply include/kernel.h]: > @pfl understood, and agreed mostly. I just think that (ignoring the standard compliance, reserved namespace issue) the _ prefix had at least a slightly better chance of screaming "private!" than z_ has.

I agree here. But as we want to be MISRA C complieant, a prefix of '_' is out of bounds anyway.

**Comment by andrewboie** [comment]
All,

We really need to make progress on this. The real issue seems to be that there is some controversy on whether the Z_ prefix is really appropriate to namespace private APIs, some feel that Z_ does not have the "don't touch this" vibe of the old _ prefix scheme.

So I think the real discussion is, "what should the new private API prefix be"?

We can discuss here or in #9882.

But to kick things off, how about "pvt_" instead of "z_" for this?

**Comment by aescolar** [comment]
> So I think the real discussion is, "what should the new private API prefix be"?

Whatever it would be, before this PR is merged, it needs to be clearly decided what is the convention for all these. Renaming APIs is an annoyance for anybody who is implementing anything which uses Zephyr, so we should not change them back and forth. If we need to, let's please do it only once.

**Comment by andrewboie** [comment]
> Renaming APIs is an annoyance for anybody who is implementing anything which uses Zephyr

What's under discussion is how to name *private* APIs. These already are deemed subject to renaming, changed semantics, or outright removal at any time.

**Comment by ceolin** [comment]
> All,
> 
> We really need to make progress on this. If we punt to 1.15, any backporting of patches to 1.14 is going to be an ordeal.
> 
> The real issue seems to be that there is some controversy on whether the Z_ prefix is really appropriate to namespace private APIs, some feel that Z_ does not have the "don't touch this" vibe of the old _ prefix scheme.
> 
> So I think the real discussion is, "what should the new private API prefix be"?
> 
> We can discuss here or in #9882.
> 
> But to kick things off, how about "pvt_" instead of "z_" for this?

The problem @jhedberg's mentioned  is that with have objects prefixed with `k_` (e.g `k_timer`). So, is it ok we use `k_` instead of using `z_` ? or rename `k_` objects to `z_` ?

I completely agree about having it before 1.14 btw.

**Comment by andyross** [comment]
Grumpy curmudgeon here:  **JUST PICK SOMETHING, THEY'RE PRIVATE APIS**.

They aren't documented.  They aren't legal to use.  No one has complained about z_ over the past year or so, which is about when it started appearing in the tree.

If there is a better choice for our private APIs to use, then we can go in and rename them all in 1.15, which will be perfectly fine because they're private APIs we can change whenever we like.  Then, in 1.16, when the "we ARE the system"[1] rule exception makes it through all the relevant international regulatory agencies, we can go back and change it to an underscore for clarity and concision.  And that will be perfectly fine too.

[1] No seriously: the reason underscores are disallowed is that they've been used historicaly in toolchain and system headers for internal symbols.   SunOS/AIX/Solaris/BSD/Linux headers don't follow this rule.  Ours needn't either, and for the same reason.  But whatever.  Just pick something.

**Comment by andrewboie** [comment]
> The problem @jhedberg's mentioned is that with have objects prefixed with k_ (e.g k_timer). So, is it ok we use k_ instead of using z_ ? or rename k_ objects to z_ ?

Definitely not.

How are you going to distinguish private vs public APIs if you name everything the same?

> If there is a better choice for our private APIs to use, then we can go in and rename them all in 1.15, which will be perfectly fine because they're private APIs we can change whenever we like. Then, in 1.16, when the "we ARE the system"[1] rule exception makes it through all the relevant international regulatory agencies, we can go back and change it to an underscore for clarity and concision. And that will be perfectly fine too.

Amen! Half the kernel is already using the z_ convention...if nobody has a credible alternative in the new few days just roll with it for 1.14 IMO

**Comment by pfl** [comment]
> Amen! Half the kernel is already using the z_ convention...if nobody has a credible alternative in the new few days just roll with it for 1.14 IMO

Time to hit the merge button then, before new conflicts arise and I have to rebase the set again.

**Review by tyrantlizard77 — APPROVED** [review]

File: subsys/net/lib/sockets/sockets.c
Comment (jukkar) [inline subsys/net/lib/sockets/sockets.c]: The naming of this function is not that great atm. This private function looks now like a global function. Imho we need to change the name to something like `wait_non_empty()` or `fifo_wait_non_empty()`. This could be done in this PR or I can send a separate PR after this one is merged.

File: subsys/net/lib/sockets/sockets_can.c
Comment (jukkar) [inline subsys/net/lib/sockets/sockets_can.c]: same here

File: subsys/net/lib/sockets/sockets_packet.c
Comment (jukkar) [inline subsys/net/lib/sockets/sockets_packet.c]: and here too. I wonder if we should place it into a library, either into global one or just local to sockets.

**Review by nashif — APPROVED** [review]

File: arch/arm/core/swap_helper.S
Comment (ioannisg) [inline arch/arm/core/swap_helper.S]: Why _arm_do_syscall is not updated?

**Review by ioannisg — COMMENTED** [review]
Looks ok. Left a comment about _arm_do_syscalls

File: arch/arm/core/swap_helper.S
Comment (pfl) [reply arch/arm/core/swap_helper.S]: It's one of those cases that exist only in .S files. The starting point for these changes was the Coccinelle script, and Coccinelle only detects code in .h and .c files. Doing a manual fixup for this.

**Comment by pfl** [comment]
Fixed ioannis' issue and rebased.

**Pull request merged**
